src/MongoDB.Driver/FieldValueSerializerHelper.cs

```diff
@@ -121,17 +116,17 @@ private static IBsonSerializer GetSerializerForValueType(IBsonSerializer fieldSe
                 return (IBsonSerializer)ienumerableSerializerConstructor.Invoke(new object[] { itemSerializer });
             }
 
-            // if the fieldSerializer is an array serializer try to adapt its itemSerializer for valueType
-            IBsonArraySerializer arraySerializer;
-            if ((arraySerializer = fieldSerializer as IBsonArraySerializer) != null)
+            if (allowScalarValueForArrayField)
             {
-                BsonSerializationInfo itemSerializationInfo;
-                if (arraySerializer.TryGetItemSerializationInfo(out itemSerializationInfo))
+                // if the fieldSerializer is an array serializer try to adapt its itemSerializer for valueType
+                IBsonArraySerializer arraySerializer;
+                if ((arraySerializer = fieldSerializer as IBsonArraySerializer) != null)
                 {
-                    if (recursionLevel == 0)
+                    BsonSerializationInfo itemSerializationInfo;
+                    if (arraySerializer.TryGetItemSerializationInfo(out itemSerializationInfo))
                     {
                         var itemSerializer = itemSerializationInfo.Serializer;
-                        return GetSerializerForValueType(itemSerializer, serializerRegistry, valueType, recursionLevel + 1);
+                        return GetSerializerForValueType(itemSerializer, serializerRegistry, valueType, allowScalarValueForArrayField: false);
```

Comment:
Allowing scalar values for array fields only is valid for one level. At the next level down the value is always false. This also prevents more than one level of recursion.